streaming/src/main/scala/org/apache/spark/streaming/scheduler/BatchInfo.scala

```diff
@@ -36,8 +38,21 @@ case class BatchInfo(
     streamIdToInputInfo: Map[Int, StreamInputInfo],
     submissionTime: Long,
     processingStartTime: Option[Long],
-    processingEndTime: Option[Long]
-  ) {
+    processingEndTime: Option[Long],
+    private[streaming] val failureReasons: Map[Int, String]) {
```

Comment:
You are exposing two public constructors now, that's not cool either. I think the best approach is to 
preserve the case class constructor, and add another private[streaming] field failureReason, which can be assigned with private[streaming] methods.